firewall_test.go

```diff
@@ -0,0 +1,109 @@
+package libnetwork
+
+import (
+	"fmt"
+	"github.com/docker/libnetwork/iptables"
+	"github.com/docker/libnetwork/netlabel"
+	"github.com/docker/libnetwork/options"
+	"strings"
+	"testing"
+)
+
+const (
+	fwdChainName = "FORWARD"
+	usrChainName = "DOCKER-USER"
+)
+
+func verifyUserChain(enabled, insert bool) error {
+	output, err := iptables.Raw("-S", fwdChainName)
+	if err != nil {
+		return err
+	}
+	rules := strings.Split(string(output), "\n")
+	if !enabled {
+		if len(rules)-1 != 1 {
+			return fmt.Errorf("chain %v: unexpected rules len=%v, %v, ",
+				fwdChainName, len(rules), rules)
+		}
+		if _, err = iptables.Raw("-S", usrChainName); err == nil {
+			return fmt.Errorf("chain %v: created unexpectedly", usrChainName)
+		}
+		return nil
+	}
+	nRules := 2
+	if insert {
+		nRules++
+	}
+	if nRules != len(rules)-1 || !strings.Contains(rules[1], usrChainName) {
+		return fmt.Errorf("chain %v: unexpected rules len=%v, %v",
+			fwdChainName, len(rules), rules)
+	}
+
+	output, err = iptables.Raw("-S", usrChainName)
+	if err != nil {
+		return err
+	}
+	rules = strings.Split(string(output), "\n")
+	if len(rules)-1 != 2 {
+		return fmt.Errorf("chain %v: unexpected rules len=%v, %v",
+			usrChainName, len(rules), rules)
+	}
+	return nil
+}
+
+func testUserChain(t *testing.T, ctrl *controller, enabled, insert bool) {
+	defer func() {
+		_, err := iptables.Raw("-F", fwdChainName)
+		if err != nil {
+			t.Fatal(err)
+		}
+		_ = iptables.RemoveExistingChain(usrChainName, "")
+	}()
+	// init. condition, FORWARD chain empty
+	// DOCKER-USER not exist
+	err := verifyUserChain(false, insert)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if insert {
+		_, err = iptables.Raw("-A", fwdChainName, "-j", "DROP")
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+	setupArrangeUserFilterRule(ctrl)
+	arrangeUserFilterRule()
+	err = verifyUserChain(enabled, insert)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestUserChain(t *testing.T) {
+	nc, err := New()
+	if err != nil {
+		t.Fatal(err)
+	}
+	c := nc.(*controller)
+	opt := options.Generic{
+		"EnableIPTables": false,
+	}
+	cfgBridge := make(map[string]interface{})
+	cfgBridge[netlabel.GenericData] = opt
+	c.cfg.Daemon.DriverCfg["bridge"] = cfgBridge
+
+	tests := []struct {
+		enable bool // enable IPTable
+		insert bool // insert other rules to FORWARD
+	}{
+		{enable: false, insert: false},
+		{enable: true, insert: false},
+		{enable: true, insert: true},
+	}
+
+	for _, i := range tests {
+		opt["EnableIPTables"] = i.enable
+		t.Logf("Test EnableIPTable=%v, insert=%v", i.enable, i.insert)
```

Comment:
Did a quick try what it would look like with my suggestions (just for the test); let me know if you think those make sense (patch here; [suggestions.patch.txt](https://github.com/docker/libnetwork/files/3820871/suggestions.patch.txt))

```go
package libnetwork

import (
	"fmt"
	"strings"
	"testing"

	"github.com/docker/libnetwork/iptables"
	"github.com/docker/libnetwork/netlabel"
	"github.com/docker/libnetwork/options"
	"gotest.tools/assert"
)

const fwdChainName = "FORWARD"

func TestUserChain(t *testing.T) {
	nc, err := New()
	assert.NilError(t, err)

	tests := []struct {
		iptables  bool
		insert    bool // insert other rules to FORWARD
		fwdChain  []string
		userChain []string
	}{
		{
			iptables: false,
			insert:   false,
			fwdChain: []string{"-P FORWARD ACCEPT"},
		},
		{
			iptables:  true,
			insert:    false,
			fwdChain:  []string{"-P FORWARD ACCEPT", "-A FORWARD -j DOCKER-USER"},
			userChain: []string{"-N DOCKER-USER", "-A DOCKER-USER -j RETURN"},
		},
		{
			iptables:  true,
			insert:    true,
			fwdChain:  []string{"-P FORWARD ACCEPT", "-A FORWARD -j DOCKER-USER", "-A FORWARD -j DROP"},
			userChain: []string{"-N DOCKER-USER", "-A DOCKER-USER -j RETURN"},
		},
	}

	for _, tc := range tests {
		tc := tc
		t.Run(fmt.Sprintf("iptables=%v,insert=%v", tc.iptables, tc.insert), func(t *testing.T) {
			c := nc.(*controller)
			c.cfg.Daemon.DriverCfg["bridge"] = map[string]interface{}{
				netlabel.GenericData: options.Generic{
					"EnableIPTables": tc.iptables,
				},
			}

			// init. condition, FORWARD chain empty DOCKER-USER not exist
			assert.DeepEqual(t, getRules(t, fwdChainName), []string{"-P FORWARD ACCEPT"})

			if tc.insert {
				_, err = iptables.Raw("-A", fwdChainName, "-j", "DROP")
				assert.NilError(t, err)
			}
			setupArrangeUserFilterRule(c)
			arrangeUserFilterRule()

			assert.DeepEqual(t, getRules(t, fwdChainName), tc.fwdChain)
			if tc.userChain != nil {
				assert.DeepEqual(t, getRules(t, userChain), tc.userChain)
			} else {
				_, err := iptables.Raw("-S", userChain)
				assert.Assert(t, err != nil, "chain %v: created unexpectedly", userChain)
			}
		})
		resetIptables(t)
	}
}

func getRules(t *testing.T, chain string) []string {
	t.Helper()
	output, err := iptables.Raw("-S", chain)
	assert.NilError(t, err, "chain %s: failed to get rules", chain)

	rules := strings.Split(string(output), "\n")
	if len(rules) > 0 {
		rules = rules[:len(rules)-1]
	}
	return rules
}

func resetIptables(t *testing.T) {
	t.Helper()
	_, err := iptables.Raw("-F", fwdChainName)
	assert.NilError(t, err)
	_ = iptables.RemoveExistingChain(userChain, "")
}

```

Reply:
this change looks fantastic to me, it is much more compact.